We prepare our data for GSEA analyses. For that purpose, we will use all the genes that are expressed and their t-value or log(foldchange).

```{r prepare-gsea_1, eval=EXPERIMENT==1, include=EXPERIMENT==1}
# We keep the t-value columns from genes_gsea
keep <- 1 # gene ID column
keep <- append(keep, grep("^t\\.", colnames(genes_gsea)))
genes_gsea <- genes_gsea[, keep]

# creamos una lista vacÃ­a
genes_list_arab_gsea <- list()
for (i in 2:ncol(genes_gsea)) {
  # orthology
  tmp <- merge(x = genes_gsea[, c(1, i)],
               y = ort,
               by = 1,
               all = FALSE)
  tmp <- tmp[, c(1, 3, 2)]
  # remove isoforms
  tmp[[2]] <-  gsub('\\.[0-9]', "", tmp[[2]])
  # # sort by ID and t-value
  # tmp <- arrange(tmp, desc(tmp[[2]]), desc(tmp[[3]]))
  # # keep one value per ID
  # tmp <- tmp[!duplicated(tmp[[2]]), ]
  # keep t-value
  tmp2 <- tmp[[3]]
  # name vector
  names(tmp2) <- tmp[[2]]
  # omit NA values
  tmp2 <- na.omit(tmp2)
  # sort in decreasing order
  tmp2 <- sort(tmp2, decreasing = TRUE)
  name <- colnames(genes_gsea)[i]
  name <- gsub("t.", "", name)
  genes_list_arab_gsea[[name]] <- tmp2
}

rm(keep, genes_gsea)
```

```{r prepare-gsea_3, eval=(exists("DO_GSEA_3") && DO_GSEA_3), include=(exists("DO_GSEA_3") && DO_GSEA_3)}
# We keep the log2(foldchange) column from genes_gsea
genes_gsea <- genes_gsea[, c(1, COLFC)]
# remove NA values
genes_gsea <- na.omit(genes_gsea)

# create empty list
genes_list_arab_gsea <- list()
tmp <- merge(x = genes_gsea,
             y = ort,
             by = 1,
             all = FALSE)
tmp <- tmp[, c(1, 3, 2)]
# remove isoforms
tmp[[2]] <- gsub('\\.[0-9]', "", tmp[[2]])
# keep log2(foldchange)
tmp2 <- tmp[[3]]
# name vector
names(tmp2) <- tmp[[2]]
# sort in decreasing order
tmp2 <- sort(tmp2, decreasing = TRUE)
genes_list_arab_gsea[[allContrasts]] <- tmp2

rm(tmp, tmp2, genes_gsea)
```

```{r save_genes_GSEA}
# save genes for GSEA enrichment arab
save_ids <- data.frame()
for (i in 1:length(genes_list_arab_gsea)){
  tmp <- as.data.frame(names(genes_list_arab_gsea[[i]]))
  colnames(tmp) <- "Gene_ID"
  tmp$condition <- tmp$Gene_ID
  tmp$condition <- gsub(".*", names(genes_list_arab_gsea[i]), tmp$condition)
  save_ids <- rbind(save_ids, tmp)
}
write.table(save_ids,
            file = paste0(SUB_DIR, "genes_for_gsea_arabidopsis.tsv"),
            quote = FALSE,
            sep = "\t",
            row.names = FALSE)
```

Gene set enrichment analyse each of the three GO hierarchies.

## Biological Process {.tabset .tabset-pills .tabset-fade}

```{r}
P_CUTOFF <- P_CUTOFF/10
```

```{r}
ONT <- "BP"
MAX_GSSIZE <-  ifelse (MAX_GSSIZE_USER == 0,
                       30*MIN_GSSIZE,
                       MAX_GSSIZE_USER)
```

```{r GSEA-BP, child='21_GSEA-enrichment.Rmd'}
```

## Molecular Function {.tabset .tabset-pills .tabset-fade}

```{r}
ONT <- "MF"
MAX_GSSIZE <-  ifelse (MAX_GSSIZE_USER == 0,
                       20*MIN_GSSIZE,
                       MAX_GSSIZE_USER)
```

```{r GSEA-MF, child='21_GSEA-enrichment.Rmd'}
```

## Cellular Component {.tabset .tabset-pills .tabset-fade}

```{r}
ONT <- "CC"
MAX_GSSIZE <-  ifelse (MAX_GSSIZE_USER == 0,
                       30*MIN_GSSIZE,
                       MAX_GSSIZE_USER)
```

```{r GSEA-CC, child='21_GSEA-enrichment.Rmd'}
```

